{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning: SQLite3 version 3.40.0 and 3.41.2 have huge performance regressions; please install version 3.41.1 or 3.42!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from owlready2 import get_ontology, Thing\n",
    "import matplotlib.pyplot as plt\n",
    "import copy\n",
    "import numpy as np\n",
    "from adjustText import adjust_text\n",
    "from googleapiclient.discovery import build\n",
    "from datetime import timedelta\n",
    "import isodate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting Data from CoDex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tutorials_url = \"../docs/supplementary/supplementary_table_6.tsv\"\n",
    "tools_url = \"../docs/supplementary/supplementary_table_2.tsv\"\n",
    "workflows_url = \"../docs/supplementary/supplementary_table_4.tsv\"\n",
    "\n",
    "tutorials_df = pd.read_csv(tutorials_url, sep=\"\\t\")\n",
    "tutorials_df = tutorials_df.drop_duplicates(keep=\"last\")\n",
    "\n",
    "tools_df = pd.read_csv(tools_url, sep=\"\\t\")\n",
    "\n",
    "workflows_df = pd.read_csv(workflows_url, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up your API key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "API_KEY = ''  # Replace with your  youtube API key\n",
    "youtube = build('youtube', 'v3', developerKey=API_KEY)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorials Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new rows, one row for each EDAM term per tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### EDAM Operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Title</th>\n",
       "      <th>Link</th>\n",
       "      <th>EDAM topic</th>\n",
       "      <th>EDAM operation</th>\n",
       "      <th>Creation</th>\n",
       "      <th>Last modification</th>\n",
       "      <th>Version</th>\n",
       "      <th>Tutorial</th>\n",
       "      <th>Slides</th>\n",
       "      <th>...</th>\n",
       "      <th>Tools</th>\n",
       "      <th>Servers with precise tool versions</th>\n",
       "      <th>Servers with tool but different versions</th>\n",
       "      <th>Feedback number</th>\n",
       "      <th>Feedback mean note</th>\n",
       "      <th>Visitors</th>\n",
       "      <th>Page views</th>\n",
       "      <th>Visit duration</th>\n",
       "      <th>Video views</th>\n",
       "      <th>EDAM operation single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence assembly visualisation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Statistical calculation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence contamination filtering</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence composition calculation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Topic                                              Title  \\\n",
       "0  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "1  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "2  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "3  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "4  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "\n",
       "                                                Link  \\\n",
       "0  https://training.galaxyproject.org//topics/ass...   \n",
       "1  https://training.galaxyproject.org//topics/ass...   \n",
       "2  https://training.galaxyproject.org//topics/ass...   \n",
       "3  https://training.galaxyproject.org//topics/ass...   \n",
       "4  https://training.galaxyproject.org//topics/ass...   \n",
       "\n",
       "                                          EDAM topic  \\\n",
       "0  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "1  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "2  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "3  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "4  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "\n",
       "                                      EDAM operation    Creation  \\\n",
       "0  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "1  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "2  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "3  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "4  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "\n",
       "  Last modification  Version Tutorial  Slides  ...  \\\n",
       "0        2025-03-20       16     True   False  ...   \n",
       "1        2025-03-20       16     True   False  ...   \n",
       "2        2025-03-20       16     True   False  ...   \n",
       "3        2025-03-20       16     True   False  ...   \n",
       "4        2025-03-20       16     True   False  ...   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "1  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "2  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "3  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "4  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "\n",
       "                  Servers with precise tool versions  \\\n",
       "0  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "1  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "2  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "3  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "4  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "\n",
       "  Servers with tool but different versions Feedback number Feedback mean note  \\\n",
       "0                                      NaN               0                NaN   \n",
       "1                                      NaN               0                NaN   \n",
       "2                                      NaN               0                NaN   \n",
       "3                                      NaN               0                NaN   \n",
       "4                                      NaN               0                NaN   \n",
       "\n",
       "   Visitors  Page views  Visit duration  Video views  \\\n",
       "0      9642       14954            3.65            0   \n",
       "1      9642       14954            3.65            0   \n",
       "2      9642       14954            3.65            0   \n",
       "3      9642       14954            3.65            0   \n",
       "4      9642       14954            3.65            0   \n",
       "\n",
       "              EDAM operation single  \n",
       "0   Sequence assembly visualisation  \n",
       "1                   Genome assembly  \n",
       "2           Statistical calculation  \n",
       "3  Sequence contamination filtering  \n",
       "4  Sequence composition calculation  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_tutorials_operation = []\n",
    "for idx, row in tutorials_df.iterrows():\n",
    "    terms = row[\"EDAM operation\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = val\n",
    "            inflated_tutorials_operation.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM operation single\"] = \"Undefined\"\n",
    "        inflated_tutorials_operation.append(row_c)\n",
    "\n",
    "df_inflated_tutorials_operation = pd.DataFrame(inflated_tutorials_operation)\n",
    "\n",
    "df_inflated_tutorials_operation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### EDAM topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Title</th>\n",
       "      <th>Link</th>\n",
       "      <th>EDAM topic</th>\n",
       "      <th>EDAM operation</th>\n",
       "      <th>Creation</th>\n",
       "      <th>Last modification</th>\n",
       "      <th>Version</th>\n",
       "      <th>Tutorial</th>\n",
       "      <th>Slides</th>\n",
       "      <th>...</th>\n",
       "      <th>Tools</th>\n",
       "      <th>Servers with precise tool versions</th>\n",
       "      <th>Servers with tool but different versions</th>\n",
       "      <th>Feedback number</th>\n",
       "      <th>Feedback mean note</th>\n",
       "      <th>Visitors</th>\n",
       "      <th>Page views</th>\n",
       "      <th>Visit duration</th>\n",
       "      <th>Video views</th>\n",
       "      <th>EDAM topic single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Sequence assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Whole genome sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Public health and epidemiology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Assembly</td>\n",
       "      <td>Genome Assembly of a bacterial genome (MRSA) s...</td>\n",
       "      <td>https://training.galaxyproject.org//topics/ass...</td>\n",
       "      <td>Sequence assembly, Whole genome sequencing, Pu...</td>\n",
       "      <td>Sequence assembly visualisation, Genome assemb...</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>2025-03-20</td>\n",
       "      <td>16</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>fastp, quast, fastqc, shovill, upload1, bandag...</td>\n",
       "      <td>UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9642</td>\n",
       "      <td>14954</td>\n",
       "      <td>3.65</td>\n",
       "      <td>0</td>\n",
       "      <td>Microbiology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Topic                                              Title  \\\n",
       "0  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "1  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "2  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "3  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "4  Assembly  Genome Assembly of a bacterial genome (MRSA) s...   \n",
       "\n",
       "                                                Link  \\\n",
       "0  https://training.galaxyproject.org//topics/ass...   \n",
       "1  https://training.galaxyproject.org//topics/ass...   \n",
       "2  https://training.galaxyproject.org//topics/ass...   \n",
       "3  https://training.galaxyproject.org//topics/ass...   \n",
       "4  https://training.galaxyproject.org//topics/ass...   \n",
       "\n",
       "                                          EDAM topic  \\\n",
       "0  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "1  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "2  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "3  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "4  Sequence assembly, Whole genome sequencing, Pu...   \n",
       "\n",
       "                                      EDAM operation    Creation  \\\n",
       "0  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "1  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "2  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "3  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "4  Sequence assembly visualisation, Genome assemb...  2021-03-24   \n",
       "\n",
       "  Last modification  Version Tutorial  Slides  ...  \\\n",
       "0        2025-03-20       16     True   False  ...   \n",
       "1        2025-03-20       16     True   False  ...   \n",
       "2        2025-03-20       16     True   False  ...   \n",
       "3        2025-03-20       16     True   False  ...   \n",
       "4        2025-03-20       16     True   False  ...   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "1  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "2  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "3  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "4  fastp, quast, fastqc, shovill, upload1, bandag...   \n",
       "\n",
       "                  Servers with precise tool versions  \\\n",
       "0  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "1  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "2  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "3  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "4  UseGalaxy.be, UseGalaxy.cz, UseGalaxy.eu, UseG...   \n",
       "\n",
       "  Servers with tool but different versions Feedback number Feedback mean note  \\\n",
       "0                                      NaN               0                NaN   \n",
       "1                                      NaN               0                NaN   \n",
       "2                                      NaN               0                NaN   \n",
       "3                                      NaN               0                NaN   \n",
       "4                                      NaN               0                NaN   \n",
       "\n",
       "   Visitors  Page views  Visit duration  Video views  \\\n",
       "0      9642       14954            3.65            0   \n",
       "1      9642       14954            3.65            0   \n",
       "2      9642       14954            3.65            0   \n",
       "3      9642       14954            3.65            0   \n",
       "4      9642       14954            3.65            0   \n",
       "\n",
       "                EDAM topic single  \n",
       "0               Sequence assembly  \n",
       "1         Whole genome sequencing  \n",
       "2  Public health and epidemiology  \n",
       "3                        Genomics  \n",
       "4                    Microbiology  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_tutorials_topics = []\n",
    "for idx, row in tutorials_df.iterrows():\n",
    "    terms = row[\"EDAM topic\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = val\n",
    "            inflated_tutorials_topics.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM topic single\"] = \"Undefined\"\n",
    "        inflated_tutorials_topics.append(row_c)\n",
    "\n",
    "df_inflated_tutorials_topics = pd.DataFrame(inflated_tutorials_topics)\n",
    "\n",
    "df_inflated_tutorials_topics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique tutorials: 39\n",
      "Total number of unique EDAM operations: 86\n",
      "Total number of unique EDAM topics: 26\n",
      "Total number of unique tutorials: 39\n",
      "Total number of tutorials with associated videos: 18\n",
      "Percentage of tutorials with videos: 46.15%\n",
      "Total number of videos: 18\n",
      "Total number of tutorials with associated workflows: 33\n",
      "Percentage of tutorials with workflows: 84.62%\n",
      "Total number of workflows: 33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_36844/3973245857.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2[\"Video Count\"] = df2[\"Video\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n",
      "/tmp/ipykernel_36844/3973245857.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2[\"Workflow Count\"] = df2[\"Workflows\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n"
     ]
    }
   ],
   "source": [
    "unique_tutorials = tutorials_df[\"Title\"].nunique()\n",
    "df2 = tutorials_df.drop_duplicates(subset='Title', keep=\"last\")\n",
    "unique_operations = df_inflated_tutorials_operation[\"EDAM operation single\"].nunique()\n",
    "unique_topics = df_inflated_tutorials_topics[\"EDAM topic single\"].nunique()\n",
    "\n",
    "# Count videos and workflows per tutorial with additional checks for string type\n",
    "df2[\"Video Count\"] = df2[\"Video\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n",
    "df2[\"Workflow Count\"] = df2[\"Workflows\"].apply(lambda x: len(x.split(',')) if isinstance(x, str) and x.strip() else 0)\n",
    "\n",
    "# Calculate tutorials with associated videos and workflows (count of True values in each column)\n",
    "tutorials_with_videos = df2[\"Video\"].sum()\n",
    "tutorials_with_workflows = df2[\"Workflows\"].sum()\n",
    "\n",
    "# Calculate the total number of videos and workflows\n",
    "total_videos = df2[df2[\"Video\"] == True].shape[0]  # Total number of True values in Video column\n",
    "total_workflows = df2[df2[\"Workflows\"] == True].shape[0]  # Total number of True values in Workflows column\n",
    "\n",
    "# Calculate percentages\n",
    "percentage_with_videos = (tutorials_with_videos / unique_tutorials) * 100\n",
    "percentage_with_workflows = (tutorials_with_workflows / unique_tutorials) * 100\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Total number of unique tutorials: {unique_tutorials}\")\n",
    "print(f\"Total number of unique EDAM operations: {unique_operations}\")\n",
    "print(f\"Total number of unique EDAM topics: {unique_topics}\")\n",
    "print(f\"Total number of unique tutorials: {unique_tutorials}\")\n",
    "print(f\"Total number of tutorials with associated videos: {tutorials_with_videos}\")\n",
    "print(f\"Percentage of tutorials with videos: {percentage_with_videos:.2f}%\")\n",
    "print(f\"Total number of videos: {total_videos}\")\n",
    "\n",
    "print(f\"Total number of tutorials with associated workflows: {tutorials_with_workflows}\")\n",
    "print(f\"Percentage of tutorials with workflows: {percentage_with_workflows:.2f}%\")\n",
    "print(f\"Total number of workflows: {total_workflows}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tutorials Videos on Youtube"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching video durations...\n",
      "\n",
      "\n",
      "Summary of Video Durations:\n",
      "Total number of unique videos: 17\n",
      "Total duration (hh:mm:ss): 15:42:09\n",
      "Updated table saved to tutorialsYoutube/table_with_api_video_durations.csv\n"
     ]
    }
   ],
   "source": [
    "# Function to get video duration using YouTube API\n",
    "def get_video_duration_api(video_url):\n",
    "    try:\n",
    "        # Extract the video ID from the URL\n",
    "        video_id = video_url.split('v=')[-1].split('&')[0] if 'youtube' in video_url else video_url.split('/')[-1]\n",
    "        request = youtube.videos().list(part=\"contentDetails\", id=video_id)\n",
    "        response = request.execute()\n",
    "\n",
    "        # Extract duration in ISO 8601 format\n",
    "        items = response.get(\"items\")\n",
    "        if items:\n",
    "            duration = items[0][\"contentDetails\"][\"duration\"]\n",
    "            # Convert ISO 8601 to total seconds\n",
    "            return parse_iso_duration(duration)\n",
    "        else:\n",
    "            print(f\"No details found for video: {video_url}\")\n",
    "            return 0\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing video {video_url}: {e}\")\n",
    "        return 0\n",
    "\n",
    "# Function to parse ISO 8601 duration to seconds\n",
    "\n",
    "def parse_iso_duration(duration):\n",
    "    duration_obj = isodate.parse_duration(duration)\n",
    "    return int(duration_obj.total_seconds())\n",
    "\n",
    "# Load the table containing YouTube links\n",
    "file_path = 'tutorialsYoutube/tutorialsVideos.csv'\n",
    "df = pd.read_csv(file_path)\n",
    "\n",
    "# Column containing YouTube links\n",
    "column_name = \"Video_Link\"\n",
    "\n",
    "# Apply function to get video durations\n",
    "print(\"Fetching video durations...\\n\")\n",
    "df['Duration_seconds'] = df[column_name].apply(get_video_duration_api)\n",
    "\n",
    "# Calculate total duration\n",
    "total_duration_seconds = int(df['Duration_seconds'].sum())  # Convert to Python int\n",
    "total_duration = timedelta(seconds=total_duration_seconds)\n",
    "\n",
    "# Display the total duration\n",
    "print(\"\\nSummary of Video Durations:\")\n",
    "print(f\"Total number of unique videos: {len(df)}\")\n",
    "print(f\"Total duration (hh:mm:ss): {total_duration}\")\n",
    "\n",
    "# Save updated table\n",
    "output_file = \"tutorialsYoutube/table_with_api_video_durations.csv\"\n",
    "df.to_csv(output_file, index=False)\n",
    "print(f\"Updated table saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tools Statistics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new rows, one row for each EDAM term per tool"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM Operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "inflated_tools = []\n",
    "\n",
    "for idx, row in tools_df.iterrows():\n",
    "    terms = row[\"EDAM reduced operations\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = val\n",
    "            inflated_tools.append(row_c)\n",
    "    else:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = \"Undefined\"\n",
    "            inflated_tools.append(row_c)\n",
    "\n",
    "df_inflated_tools_operation = pd.DataFrame(inflated_tools)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM Topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "inflated_tools = []\n",
    "\n",
    "for idx, row in tools_df.iterrows():\n",
    "    terms = row[\"EDAM reduced topics\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = val\n",
    "            inflated_tools.append(row_c)\n",
    "    else:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = \"Undefined\"\n",
    "            inflated_tools.append(row_c)\n",
    "\n",
    "df_inflated_tools_topics = pd.DataFrame(inflated_tools)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tools not suites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Suite ID</th>\n",
       "      <th>Tool IDs</th>\n",
       "      <th>Description</th>\n",
       "      <th>Suite first commit date</th>\n",
       "      <th>Homepage</th>\n",
       "      <th>Suite version</th>\n",
       "      <th>Suite conda package</th>\n",
       "      <th>Latest suite conda package version</th>\n",
       "      <th>Suite version status</th>\n",
       "      <th>ToolShed categories</th>\n",
       "      <th>...</th>\n",
       "      <th>Suite runs (usegalaxy.fr)</th>\n",
       "      <th>Suite runs on main servers</th>\n",
       "      <th>Suite runs (last 5 years) on main servers</th>\n",
       "      <th>Suite users on main servers</th>\n",
       "      <th>Suite users (last 5 years) on main servers</th>\n",
       "      <th>Related Workflows</th>\n",
       "      <th>Related Tutorials</th>\n",
       "      <th>To keep</th>\n",
       "      <th>Deprecated</th>\n",
       "      <th>tools per suite</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMRFinderPlus</td>\n",
       "      <td>amrfinderplus</td>\n",
       "      <td>\"AMRFinderPlus is designed to find acquired an...</td>\n",
       "      <td>2023-05-12</td>\n",
       "      <td>https://github.com/ncbi/amr</td>\n",
       "      <td>3.12.8</td>\n",
       "      <td>ncbi-amrfinderplus</td>\n",
       "      <td>4.0.23</td>\n",
       "      <td>To update</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>1863</td>\n",
       "      <td>64355</td>\n",
       "      <td>64355</td>\n",
       "      <td>1670</td>\n",
       "      <td>1670</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=1ce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>amrfinderplus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISEScan</td>\n",
       "      <td>isescan</td>\n",
       "      <td>\"ISEScan is a pipeline to identify IS (Inserti...</td>\n",
       "      <td>2022-09-01</td>\n",
       "      <td>https://github.com/xiezhq/ISEScan</td>\n",
       "      <td>1.7.3</td>\n",
       "      <td>isescan</td>\n",
       "      <td>1.7.3</td>\n",
       "      <td>Up-to-date</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>1759</td>\n",
       "      <td>78274</td>\n",
       "      <td>78274</td>\n",
       "      <td>782</td>\n",
       "      <td>782</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=2c1...</td>\n",
       "      <td>genome-annotation/bacterial-genome-annotation</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>isescan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>abacas</td>\n",
       "      <td>abacas</td>\n",
       "      <td>Order and Orientate Contigs</td>\n",
       "      <td>2019-11-20</td>\n",
       "      <td>https://github.com/phac-nml/abacas</td>\n",
       "      <td>1.1</td>\n",
       "      <td>mummer</td>\n",
       "      <td>3.23</td>\n",
       "      <td>To update</td>\n",
       "      <td>Assembly</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>abacas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>abricate</td>\n",
       "      <td>abricate, abricate_list, abricate_summary</td>\n",
       "      <td>Mass screening of contigs for antiobiotic resi...</td>\n",
       "      <td>2016-07-29</td>\n",
       "      <td>https://github.com/tseemann/abricate</td>\n",
       "      <td>1.0.1</td>\n",
       "      <td>abricate</td>\n",
       "      <td>1.0.1</td>\n",
       "      <td>Up-to-date</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>19178</td>\n",
       "      <td>1846906</td>\n",
       "      <td>1798743</td>\n",
       "      <td>10448</td>\n",
       "      <td>9875</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=1ce...</td>\n",
       "      <td>microbiome/pathogen-detection-from-nanopore-fo...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>abricate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>abricate</td>\n",
       "      <td>abricate, abricate_list, abricate_summary</td>\n",
       "      <td>Mass screening of contigs for antiobiotic resi...</td>\n",
       "      <td>2016-07-29</td>\n",
       "      <td>https://github.com/tseemann/abricate</td>\n",
       "      <td>1.0.1</td>\n",
       "      <td>abricate</td>\n",
       "      <td>1.0.1</td>\n",
       "      <td>Up-to-date</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>19178</td>\n",
       "      <td>1846906</td>\n",
       "      <td>1798743</td>\n",
       "      <td>10448</td>\n",
       "      <td>9875</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=1ce...</td>\n",
       "      <td>microbiome/pathogen-detection-from-nanopore-fo...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>abricate_list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>vsearch</td>\n",
       "      <td>vsearch_alignment, vsearch_chimera_detection, ...</td>\n",
       "      <td>VSEARCH including searching, clustering, chime...</td>\n",
       "      <td>2014-11-30</td>\n",
       "      <td>https://github.com/torognes/vsearch</td>\n",
       "      <td>2.8.3</td>\n",
       "      <td>vsearch</td>\n",
       "      <td>2.30.0</td>\n",
       "      <td>To update</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>286</td>\n",
       "      <td>83368</td>\n",
       "      <td>35354</td>\n",
       "      <td>1010</td>\n",
       "      <td>642</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=401...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>vsearch_masking</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>vsearch</td>\n",
       "      <td>vsearch_alignment, vsearch_chimera_detection, ...</td>\n",
       "      <td>VSEARCH including searching, clustering, chime...</td>\n",
       "      <td>2014-11-30</td>\n",
       "      <td>https://github.com/torognes/vsearch</td>\n",
       "      <td>2.8.3</td>\n",
       "      <td>vsearch</td>\n",
       "      <td>2.30.0</td>\n",
       "      <td>To update</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>286</td>\n",
       "      <td>83368</td>\n",
       "      <td>35354</td>\n",
       "      <td>1010</td>\n",
       "      <td>642</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=401...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>vsearch_search</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910</th>\n",
       "      <td>vsearch</td>\n",
       "      <td>vsearch_alignment, vsearch_chimera_detection, ...</td>\n",
       "      <td>VSEARCH including searching, clustering, chime...</td>\n",
       "      <td>2014-11-30</td>\n",
       "      <td>https://github.com/torognes/vsearch</td>\n",
       "      <td>2.8.3</td>\n",
       "      <td>vsearch</td>\n",
       "      <td>2.30.0</td>\n",
       "      <td>To update</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>286</td>\n",
       "      <td>83368</td>\n",
       "      <td>35354</td>\n",
       "      <td>1010</td>\n",
       "      <td>642</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=401...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>vsearch_shuffling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911</th>\n",
       "      <td>vsearch</td>\n",
       "      <td>vsearch_alignment, vsearch_chimera_detection, ...</td>\n",
       "      <td>VSEARCH including searching, clustering, chime...</td>\n",
       "      <td>2014-11-30</td>\n",
       "      <td>https://github.com/torognes/vsearch</td>\n",
       "      <td>2.8.3</td>\n",
       "      <td>vsearch</td>\n",
       "      <td>2.30.0</td>\n",
       "      <td>To update</td>\n",
       "      <td>Sequence Analysis</td>\n",
       "      <td>...</td>\n",
       "      <td>286</td>\n",
       "      <td>83368</td>\n",
       "      <td>35354</td>\n",
       "      <td>1010</td>\n",
       "      <td>642</td>\n",
       "      <td>https://usegalaxy.eu/published/workflow?id=401...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>vsearch_sorting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912</th>\n",
       "      <td>wtdbg</td>\n",
       "      <td>wtdbg</td>\n",
       "      <td>WTDBG is a fuzzy Bruijn graph (FBG) approach t...</td>\n",
       "      <td>2018-06-12</td>\n",
       "      <td>https://github.com/ruanjue/wtdbg2</td>\n",
       "      <td>2.5</td>\n",
       "      <td>wtdbg</td>\n",
       "      <td>2.5</td>\n",
       "      <td>Up-to-date</td>\n",
       "      <td>Assembly</td>\n",
       "      <td>...</td>\n",
       "      <td>145</td>\n",
       "      <td>2049</td>\n",
       "      <td>1458</td>\n",
       "      <td>479</td>\n",
       "      <td>394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>wtdbg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>913 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Suite ID                                           Tool IDs  \\\n",
       "0    AMRFinderPlus                                      amrfinderplus   \n",
       "1          ISEScan                                            isescan   \n",
       "2           abacas                                             abacas   \n",
       "3         abricate          abricate, abricate_list, abricate_summary   \n",
       "4         abricate          abricate, abricate_list, abricate_summary   \n",
       "..             ...                                                ...   \n",
       "908        vsearch  vsearch_alignment, vsearch_chimera_detection, ...   \n",
       "909        vsearch  vsearch_alignment, vsearch_chimera_detection, ...   \n",
       "910        vsearch  vsearch_alignment, vsearch_chimera_detection, ...   \n",
       "911        vsearch  vsearch_alignment, vsearch_chimera_detection, ...   \n",
       "912          wtdbg                                              wtdbg   \n",
       "\n",
       "                                           Description  \\\n",
       "0    \"AMRFinderPlus is designed to find acquired an...   \n",
       "1    \"ISEScan is a pipeline to identify IS (Inserti...   \n",
       "2                          Order and Orientate Contigs   \n",
       "3    Mass screening of contigs for antiobiotic resi...   \n",
       "4    Mass screening of contigs for antiobiotic resi...   \n",
       "..                                                 ...   \n",
       "908  VSEARCH including searching, clustering, chime...   \n",
       "909  VSEARCH including searching, clustering, chime...   \n",
       "910  VSEARCH including searching, clustering, chime...   \n",
       "911  VSEARCH including searching, clustering, chime...   \n",
       "912  WTDBG is a fuzzy Bruijn graph (FBG) approach t...   \n",
       "\n",
       "    Suite first commit date                              Homepage  \\\n",
       "0                2023-05-12           https://github.com/ncbi/amr   \n",
       "1                2022-09-01     https://github.com/xiezhq/ISEScan   \n",
       "2                2019-11-20    https://github.com/phac-nml/abacas   \n",
       "3                2016-07-29  https://github.com/tseemann/abricate   \n",
       "4                2016-07-29  https://github.com/tseemann/abricate   \n",
       "..                      ...                                   ...   \n",
       "908              2014-11-30   https://github.com/torognes/vsearch   \n",
       "909              2014-11-30   https://github.com/torognes/vsearch   \n",
       "910              2014-11-30   https://github.com/torognes/vsearch   \n",
       "911              2014-11-30   https://github.com/torognes/vsearch   \n",
       "912              2018-06-12     https://github.com/ruanjue/wtdbg2   \n",
       "\n",
       "    Suite version Suite conda package Latest suite conda package version  \\\n",
       "0          3.12.8  ncbi-amrfinderplus                             4.0.23   \n",
       "1           1.7.3             isescan                              1.7.3   \n",
       "2             1.1              mummer                               3.23   \n",
       "3           1.0.1            abricate                              1.0.1   \n",
       "4           1.0.1            abricate                              1.0.1   \n",
       "..            ...                 ...                                ...   \n",
       "908         2.8.3             vsearch                             2.30.0   \n",
       "909         2.8.3             vsearch                             2.30.0   \n",
       "910         2.8.3             vsearch                             2.30.0   \n",
       "911         2.8.3             vsearch                             2.30.0   \n",
       "912           2.5               wtdbg                                2.5   \n",
       "\n",
       "    Suite version status ToolShed categories  ... Suite runs (usegalaxy.fr)  \\\n",
       "0              To update   Sequence Analysis  ...                      1863   \n",
       "1             Up-to-date   Sequence Analysis  ...                      1759   \n",
       "2              To update            Assembly  ...                         3   \n",
       "3             Up-to-date   Sequence Analysis  ...                     19178   \n",
       "4             Up-to-date   Sequence Analysis  ...                     19178   \n",
       "..                   ...                 ...  ...                       ...   \n",
       "908            To update   Sequence Analysis  ...                       286   \n",
       "909            To update   Sequence Analysis  ...                       286   \n",
       "910            To update   Sequence Analysis  ...                       286   \n",
       "911            To update   Sequence Analysis  ...                       286   \n",
       "912           Up-to-date            Assembly  ...                       145   \n",
       "\n",
       "    Suite runs on main servers Suite runs (last 5 years) on main servers  \\\n",
       "0                        64355                                     64355   \n",
       "1                        78274                                     78274   \n",
       "2                            3                                         3   \n",
       "3                      1846906                                   1798743   \n",
       "4                      1846906                                   1798743   \n",
       "..                         ...                                       ...   \n",
       "908                      83368                                     35354   \n",
       "909                      83368                                     35354   \n",
       "910                      83368                                     35354   \n",
       "911                      83368                                     35354   \n",
       "912                       2049                                      1458   \n",
       "\n",
       "    Suite users on main servers Suite users (last 5 years) on main servers  \\\n",
       "0                          1670                                       1670   \n",
       "1                           782                                        782   \n",
       "2                             2                                          2   \n",
       "3                         10448                                       9875   \n",
       "4                         10448                                       9875   \n",
       "..                          ...                                        ...   \n",
       "908                        1010                                        642   \n",
       "909                        1010                                        642   \n",
       "910                        1010                                        642   \n",
       "911                        1010                                        642   \n",
       "912                         479                                        394   \n",
       "\n",
       "                                     Related Workflows  \\\n",
       "0    https://usegalaxy.eu/published/workflow?id=1ce...   \n",
       "1    https://usegalaxy.eu/published/workflow?id=2c1...   \n",
       "2                                                  NaN   \n",
       "3    https://usegalaxy.eu/published/workflow?id=1ce...   \n",
       "4    https://usegalaxy.eu/published/workflow?id=1ce...   \n",
       "..                                                 ...   \n",
       "908  https://usegalaxy.eu/published/workflow?id=401...   \n",
       "909  https://usegalaxy.eu/published/workflow?id=401...   \n",
       "910  https://usegalaxy.eu/published/workflow?id=401...   \n",
       "911  https://usegalaxy.eu/published/workflow?id=401...   \n",
       "912                                                NaN   \n",
       "\n",
       "                                     Related Tutorials To keep Deprecated  \\\n",
       "0                                                  NaN    True      False   \n",
       "1        genome-annotation/bacterial-genome-annotation    True      False   \n",
       "2                                                  NaN    True      False   \n",
       "3    microbiome/pathogen-detection-from-nanopore-fo...    True      False   \n",
       "4    microbiome/pathogen-detection-from-nanopore-fo...    True      False   \n",
       "..                                                 ...     ...        ...   \n",
       "908                                                NaN    True      False   \n",
       "909                                                NaN    True      False   \n",
       "910                                                NaN    True      False   \n",
       "911                                                NaN    True      False   \n",
       "912                                                NaN    True      False   \n",
       "\n",
       "       tools per suite  \n",
       "0        amrfinderplus  \n",
       "1              isescan  \n",
       "2               abacas  \n",
       "3             abricate  \n",
       "4        abricate_list  \n",
       "..                 ...  \n",
       "908    vsearch_masking  \n",
       "909     vsearch_search  \n",
       "910  vsearch_shuffling  \n",
       "911    vsearch_sorting  \n",
       "912              wtdbg  \n",
       "\n",
       "[913 rows x 75 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_tools = []\n",
    "\n",
    "for idx, row in tools_df.iterrows():\n",
    "    terms = row[\"Tool IDs\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"tools per suite\"] = val\n",
    "            inflated_tools.append(row_c)\n",
    "    else:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"tools per suite\"] = \"Undefined\"\n",
    "            inflated_tools.append(row_c)\n",
    "\n",
    "df_inflated_tools_per_suite = pd.DataFrame(inflated_tools)\n",
    "df_inflated_tools_per_suite"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(320,\n",
       " 25164728,\n",
       " 766510,\n",
       " 2395.34375,\n",
       " Suite ID                                                                                  fastp\n",
       " Tool IDs                                                                                  fastp\n",
       " Description                                       Fast all-in-one preprocessing for FASTQ files\n",
       " Suite first commit date                                                              2018-03-07\n",
       " Homepage                                                      https://github.com/OpenGene/fastp\n",
       "                                                                     ...                        \n",
       " Suite users (last 5 years) on main servers                                                27815\n",
       " Related Workflows                             https://usegalaxy.eu/published/workflow?id=007...\n",
       " Related Tutorials                             assembly/assembly-with-preprocessing, assembly...\n",
       " To keep                                                                                    True\n",
       " Deprecated                                                                                False\n",
       " Name: 84, Length: 74, dtype: object,\n",
       " Suite ID                                                                                multiqc\n",
       " Tool IDs                                                                                multiqc\n",
       " Description                                   MultiQC aggregates results from bioinformatics...\n",
       " Suite first commit date                                                              2017-09-21\n",
       " Homepage                                                                   http://multiqc.info/\n",
       "                                                                     ...                        \n",
       " Suite users (last 5 years) on main servers                                                65767\n",
       " Related Workflows                             https://usegalaxy.eu/published/workflow?id=002...\n",
       " Related Tutorials                             assembly/assembly-with-preprocessing, assembly...\n",
       " To keep                                                                                    True\n",
       " Deprecated                                                                                False\n",
       " Name: 206, Length: 74, dtype: object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the total number of tools\n",
    "total_tools = tools_df.shape[0]\n",
    "\n",
    "# Calculate total tool usage over the last 5 years, over all main servers\n",
    "total_tool_usage = tools_df['Suite runs (last 5 years) on main servers'].sum()\n",
    "\n",
    "# Calculate total number of tool users over the last 5 years, over all main servers\n",
    "total_tool_users = tools_df['Suite users (last 5 years) on main servers'].sum()\n",
    "\n",
    "# Calculate the average number of tool users over the last 5 years, over all main servers\n",
    "average_tool_users = tools_df['Suite users (last 5 years) on main servers'].mean()\n",
    "\n",
    "# Find the most used tool over the last 5 years, over all main servers\n",
    "most_used_tool = tools_df.loc[tools_df['Suite runs (last 5 years) on main servers'].idxmax()]\n",
    "\n",
    "# Find the tool with the most used over the last 5 years, over all main servers\n",
    "most_users_tool = tools_df.loc[tools_df['Suite users (last 5 years) on main servers'].idxmax()]\n",
    "\n",
    "total_tools, total_tool_usage, total_tool_users, average_tool_users, most_used_tool, most_users_tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of tools: 318\n",
      "Total number of unique EDAM operations: 155\n",
      "Total number of unique EDAM topics: 97\n",
      "Total number of unique tools: 910\n"
     ]
    }
   ],
   "source": [
    "unique_tools = tools_df[\"Suite ID\"].nunique()\n",
    "unique_tools_operations = df_inflated_tools_operation[\"EDAM operation single\"].nunique()\n",
    "unique_tools_topics = df_inflated_tools_topics[\"EDAM topic single\"].nunique()\n",
    "unique_tools_per_suite = df_inflated_tools_per_suite[\"tools per suite\"].nunique()\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Total number of tools: {unique_tools}\")\n",
    "print(f\"Total number of unique EDAM operations: {unique_tools_operations}\")\n",
    "print(f\"Total number of unique EDAM topics: {unique_tools_topics}\")\n",
    "print(f\"Total number of unique tools: {unique_tools_per_suite}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflows Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create new rows, one row for each EDAM term per workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM Operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Source</th>\n",
       "      <th>ID</th>\n",
       "      <th>Link</th>\n",
       "      <th>Creators</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Creation time</th>\n",
       "      <th>Update time</th>\n",
       "      <th>Latest version</th>\n",
       "      <th>Versions</th>\n",
       "      <th>Number of steps</th>\n",
       "      <th>Tools</th>\n",
       "      <th>EDAM operations</th>\n",
       "      <th>EDAM topics</th>\n",
       "      <th>License</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Projects</th>\n",
       "      <th>To keep</th>\n",
       "      <th>Deprecated</th>\n",
       "      <th>EDAM operation single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacterial-quality-and-contamination-control-po...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1882</td>\n",
       "      <td>https://workflowhub.eu/workflows/1882?version=2</td>\n",
       "      <td>ABRomics None, Pierre Marin, Clea Siguret</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>est_abundance, kraken2, checkm2, quast, collap...</td>\n",
       "      <td>Cross-assembly, Taxonomic classification, Stat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Cross-assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacterial-quality-and-contamination-control-po...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1882</td>\n",
       "      <td>https://workflowhub.eu/workflows/1882?version=2</td>\n",
       "      <td>ABRomics None, Pierre Marin, Clea Siguret</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>est_abundance, kraken2, checkm2, quast, collap...</td>\n",
       "      <td>Cross-assembly, Taxonomic classification, Stat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Taxonomic classification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bacterial-quality-and-contamination-control-po...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1882</td>\n",
       "      <td>https://workflowhub.eu/workflows/1882?version=2</td>\n",
       "      <td>ABRomics None, Pierre Marin, Clea Siguret</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>est_abundance, kraken2, checkm2, quast, collap...</td>\n",
       "      <td>Cross-assembly, Taxonomic classification, Stat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Statistical calculation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bacterial-quality-and-contamination-control-po...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1882</td>\n",
       "      <td>https://workflowhub.eu/workflows/1882?version=2</td>\n",
       "      <td>ABRomics None, Pierre Marin, Clea Siguret</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>est_abundance, kraken2, checkm2, quast, collap...</td>\n",
       "      <td>Cross-assembly, Taxonomic classification, Stat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Expression analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bacterial-quality-and-contamination-control-po...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1882</td>\n",
       "      <td>https://workflowhub.eu/workflows/1882?version=2</td>\n",
       "      <td>ABRomics None, Pierre Marin, Clea Siguret</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>est_abundance, kraken2, checkm2, quast, collap...</td>\n",
       "      <td>Cross-assembly, Taxonomic classification, Stat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Visualisation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Name       Source    ID  \\\n",
       "0  bacterial-quality-and-contamination-control-po...  WorkflowHub  1882   \n",
       "1  bacterial-quality-and-contamination-control-po...  WorkflowHub  1882   \n",
       "2  bacterial-quality-and-contamination-control-po...  WorkflowHub  1882   \n",
       "3  bacterial-quality-and-contamination-control-po...  WorkflowHub  1882   \n",
       "4  bacterial-quality-and-contamination-control-po...  WorkflowHub  1882   \n",
       "\n",
       "                                              Link  \\\n",
       "0  https://workflowhub.eu/workflows/1882?version=2   \n",
       "1  https://workflowhub.eu/workflows/1882?version=2   \n",
       "2  https://workflowhub.eu/workflows/1882?version=2   \n",
       "3  https://workflowhub.eu/workflows/1882?version=2   \n",
       "4  https://workflowhub.eu/workflows/1882?version=2   \n",
       "\n",
       "                                    Creators Tags Creation time Update time  \\\n",
       "0  ABRomics None, Pierre Marin, Clea Siguret  NaN    2025-08-28  2025-08-28   \n",
       "1  ABRomics None, Pierre Marin, Clea Siguret  NaN    2025-08-28  2025-08-28   \n",
       "2  ABRomics None, Pierre Marin, Clea Siguret  NaN    2025-08-28  2025-08-28   \n",
       "3  ABRomics None, Pierre Marin, Clea Siguret  NaN    2025-08-28  2025-08-28   \n",
       "4  ABRomics None, Pierre Marin, Clea Siguret  NaN    2025-08-28  2025-08-28   \n",
       "\n",
       "   Latest version  Versions  Number of steps  \\\n",
       "0               2         2             14.0   \n",
       "1               2         2             14.0   \n",
       "2               2         2             14.0   \n",
       "3               2         2             14.0   \n",
       "4               2         2             14.0   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  est_abundance, kraken2, checkm2, quast, collap...   \n",
       "1  est_abundance, kraken2, checkm2, quast, collap...   \n",
       "2  est_abundance, kraken2, checkm2, quast, collap...   \n",
       "3  est_abundance, kraken2, checkm2, quast, collap...   \n",
       "4  est_abundance, kraken2, checkm2, quast, collap...   \n",
       "\n",
       "                                     EDAM operations EDAM topics  \\\n",
       "0  Cross-assembly, Taxonomic classification, Stat...         NaN   \n",
       "1  Cross-assembly, Taxonomic classification, Stat...         NaN   \n",
       "2  Cross-assembly, Taxonomic classification, Stat...         NaN   \n",
       "3  Cross-assembly, Taxonomic classification, Stat...         NaN   \n",
       "4  Cross-assembly, Taxonomic classification, Stat...         NaN   \n",
       "\n",
       "            License  DOI                                 Projects  To keep  \\\n",
       "0  GPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "1  GPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "2  GPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "3  GPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "4  GPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "\n",
       "   Deprecated     EDAM operation single  \n",
       "0       False            Cross-assembly  \n",
       "1       False  Taxonomic classification  \n",
       "2       False   Statistical calculation  \n",
       "3       False       Expression analysis  \n",
       "4       False             Visualisation  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_workflows_operation = []\n",
    "for idx, row in workflows_df.iterrows():\n",
    "    terms = row[\"EDAM operations\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM operation single\"] = val\n",
    "            inflated_workflows_operation.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM operation single\"] = \"Undefined\"\n",
    "        inflated_workflows_operation.append(row_c)\n",
    "\n",
    "df_inflated_workflows_operation = pd.DataFrame(inflated_workflows_operation)\n",
    "\n",
    "df_inflated_workflows_operation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EDAM topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Source</th>\n",
       "      <th>ID</th>\n",
       "      <th>Link</th>\n",
       "      <th>Creators</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Creation time</th>\n",
       "      <th>Update time</th>\n",
       "      <th>Latest version</th>\n",
       "      <th>Versions</th>\n",
       "      <th>Number of steps</th>\n",
       "      <th>Tools</th>\n",
       "      <th>EDAM operations</th>\n",
       "      <th>EDAM topics</th>\n",
       "      <th>License</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Projects</th>\n",
       "      <th>To keep</th>\n",
       "      <th>Deprecated</th>\n",
       "      <th>EDAM topic single</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacterial-quality-and-contamination-control-po...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1882</td>\n",
       "      <td>https://workflowhub.eu/workflows/1882?version=2</td>\n",
       "      <td>ABRomics None, Pierre Marin, Clea Siguret</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2025-08-28</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>est_abundance, kraken2, checkm2, quast, collap...</td>\n",
       "      <td>Cross-assembly, Taxonomic classification, Stat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sars-cov-2-pe-illumina-artic-ivar-analysis/SAR...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>155</td>\n",
       "      <td>https://workflowhub.eu/workflows/155?version=6</td>\n",
       "      <td>Peter van Heusden</td>\n",
       "      <td>virology</td>\n",
       "      <td>2025-07-31</td>\n",
       "      <td>2025-08-19</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>16.0</td>\n",
       "      <td>tp_cat, multiqc, ivar_variants, snpeff_sars_co...</td>\n",
       "      <td>Generation, Sequence alignment, Variant classi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sars-cov-2-pe-illumina-wgs-variant-calling/COV...</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>113</td>\n",
       "      <td>https://workflowhub.eu/workflows/113?version=4</td>\n",
       "      <td>Wolfgang Maier</td>\n",
       "      <td>virology, covid-19, covid19.galaxyproject.org,...</td>\n",
       "      <td>2021-12-21</td>\n",
       "      <td>2025-08-19</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>11.0</td>\n",
       "      <td>multiqc, snpeff_sars_cov_2, picard_MarkDuplica...</td>\n",
       "      <td>Generation, SNP detection, Genome indexing, Re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>generic-non-segmented-viral-variant-calling/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1876</td>\n",
       "      <td>https://workflowhub.eu/workflows/1876?version=1</td>\n",
       "      <td>Peter van Heusden, Wolfgang Maier</td>\n",
       "      <td>virology</td>\n",
       "      <td>2025-08-19</td>\n",
       "      <td>2025-08-19</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>23.0</td>\n",
       "      <td>qualimap_bamqc, samtools_view, snpEff, tp_cat,...</td>\n",
       "      <td>Generation, Genome indexing, Read mapping, Val...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>influenza-isolates-consensus-and-subtyping/main</td>\n",
       "      <td>WorkflowHub</td>\n",
       "      <td>1260</td>\n",
       "      <td>https://workflowhub.eu/workflows/1260?version=2</td>\n",
       "      <td>Viktoria Isabel Schwarz, Wolfgang Maier, Aaron...</td>\n",
       "      <td>virology</td>\n",
       "      <td>2025-05-14</td>\n",
       "      <td>2025-08-19</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>41.0</td>\n",
       "      <td>seqtk_subseq, qualimap_bamqc, samtools_view, _...</td>\n",
       "      <td>Sequence file editing, Generation, Primer remo...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGPL-3.0-or-later</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intergalactic Workflow Commission (IWC)</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Name       Source    ID  \\\n",
       "0  bacterial-quality-and-contamination-control-po...  WorkflowHub  1882   \n",
       "1  sars-cov-2-pe-illumina-artic-ivar-analysis/SAR...  WorkflowHub   155   \n",
       "2  sars-cov-2-pe-illumina-wgs-variant-calling/COV...  WorkflowHub   113   \n",
       "3   generic-non-segmented-viral-variant-calling/main  WorkflowHub  1876   \n",
       "4    influenza-isolates-consensus-and-subtyping/main  WorkflowHub  1260   \n",
       "\n",
       "                                              Link  \\\n",
       "0  https://workflowhub.eu/workflows/1882?version=2   \n",
       "1   https://workflowhub.eu/workflows/155?version=6   \n",
       "2   https://workflowhub.eu/workflows/113?version=4   \n",
       "3  https://workflowhub.eu/workflows/1876?version=1   \n",
       "4  https://workflowhub.eu/workflows/1260?version=2   \n",
       "\n",
       "                                            Creators  \\\n",
       "0          ABRomics None, Pierre Marin, Clea Siguret   \n",
       "1                                  Peter van Heusden   \n",
       "2                                     Wolfgang Maier   \n",
       "3                  Peter van Heusden, Wolfgang Maier   \n",
       "4  Viktoria Isabel Schwarz, Wolfgang Maier, Aaron...   \n",
       "\n",
       "                                                Tags Creation time  \\\n",
       "0                                                NaN    2025-08-28   \n",
       "1                                           virology    2025-07-31   \n",
       "2  virology, covid-19, covid19.galaxyproject.org,...    2021-12-21   \n",
       "3                                           virology    2025-08-19   \n",
       "4                                           virology    2025-05-14   \n",
       "\n",
       "  Update time  Latest version  Versions  Number of steps  \\\n",
       "0  2025-08-28               2         2             14.0   \n",
       "1  2025-08-19               6         6             16.0   \n",
       "2  2025-08-19               4         4             11.0   \n",
       "3  2025-08-19               1         1             23.0   \n",
       "4  2025-08-19               2         2             41.0   \n",
       "\n",
       "                                               Tools  \\\n",
       "0  est_abundance, kraken2, checkm2, quast, collap...   \n",
       "1  tp_cat, multiqc, ivar_variants, snpeff_sars_co...   \n",
       "2  multiqc, snpeff_sars_cov_2, picard_MarkDuplica...   \n",
       "3  qualimap_bamqc, samtools_view, snpEff, tp_cat,...   \n",
       "4  seqtk_subseq, qualimap_bamqc, samtools_view, _...   \n",
       "\n",
       "                                     EDAM operations EDAM topics  \\\n",
       "0  Cross-assembly, Taxonomic classification, Stat...         NaN   \n",
       "1  Generation, Sequence alignment, Variant classi...         NaN   \n",
       "2  Generation, SNP detection, Genome indexing, Re...         NaN   \n",
       "3  Generation, Genome indexing, Read mapping, Val...         NaN   \n",
       "4  Sequence file editing, Generation, Primer remo...         NaN   \n",
       "\n",
       "             License  DOI                                 Projects  To keep  \\\n",
       "0   GPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "1                MIT  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "2                MIT  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "3  AGPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "4  AGPL-3.0-or-later  NaN  Intergalactic Workflow Commission (IWC)     True   \n",
       "\n",
       "   Deprecated EDAM topic single  \n",
       "0       False         Undefined  \n",
       "1       False         Undefined  \n",
       "2       False         Undefined  \n",
       "3       False         Undefined  \n",
       "4       False         Undefined  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inflated_workflows_topics = []\n",
    "for idx, row in workflows_df.iterrows():\n",
    "    terms = row[\"EDAM topics\"]\n",
    "    row_d = row.to_dict()\n",
    "    if isinstance(terms, str):\n",
    "        vals = [t.strip() for t in terms.split(\",\")]\n",
    "        for val in vals:\n",
    "            row_c = copy.copy(row_d)\n",
    "            row_c[\"EDAM topic single\"] = val\n",
    "            inflated_workflows_topics.append(row_c)\n",
    "    else:\n",
    "        row_c = copy.copy(row_d)\n",
    "        row_c[\"EDAM topic single\"] = \"Undefined\"\n",
    "        inflated_workflows_topics.append(row_c)\n",
    "\n",
    "df_inflated_workflows_topics = pd.DataFrame(inflated_workflows_topics)\n",
    "\n",
    "df_inflated_workflows_topics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of workflows: 117\n",
      "Total number of unique EDAM operations: 115\n",
      "Total number of unique EDAM topics: 9\n"
     ]
    }
   ],
   "source": [
    "unique_workflows = workflows_df[\"Name\"].nunique()\n",
    "unique_workflows_operations = df_inflated_workflows_operation[\"EDAM operation single\"].nunique()\n",
    "unique_workflows_topics = df_inflated_workflows_topics[\"EDAM topic single\"].nunique()\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Total number of workflows: {unique_workflows}\")\n",
    "print(f\"Total number of unique EDAM operations: {unique_workflows_operations}\")\n",
    "print(f\"Total number of unique EDAM topics: {unique_workflows_topics}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
